src/base/misc/fatfs.c

```diff
@@ -1572,7 +1574,14 @@ void build_boot_blk(fatfs_t *f, unsigned char *b)
     case OLDMSD_D:
       txfr->ofs = 0x0000;
       txfr->seg = 0x0070;
-      txfr->ax  = d_o >> 16;
+      /*
+       * MS-DOS 3.10 needs the offset of MSDOS.SYS / 16 passed in AX and other
+       * versions don't seem to mind. See the issue discussion at
+       * https://github.com/stsp/dosemu2/issues/278. Note this relies upon the
+       * ordering of the directory, as was a requirement with earlier DOSes.
+       */
+      clstrsize = f->cluster_secs * SECTOR_SIZE;
+      txfr->ax  = ((f->obj[1].size + clstrsize - 1) & ~clstrsize) / 16;
```

Comment:
Okay I tried to set it to
```C
txfr->ax  = ((f->obj[2].start - 2) * f->cluster_secs * SECTOR_SIZE) / 16;
````
but it seems the filesystem hasn't been fully created at this point, e.g. in the log
~~~
DIR: Selected floppy disk type (5 1/4 inch 360Kb)                               
DIR auto disk /clients/common/dosemu2.git/test-imagedir/drives/c; h=2, s=9, t=40, start=0
fatfs: init: /clients/common/dosemu2.git/test-imagedir/drives/c                 
fatfs: new_obj: created obj 0 (1 left)                                          
fatfs: init: volume label set to "DRIVES C   "                                  
fatfs: scan_dir: reading "/clients/common/dosemu2.git/test-imagedir/drives/c"   
fatfs: full_name: 0 = "/clients/common/dosemu2.git/test-imagedir/drives/c/"     
fatfs: full_name: 0 = "/clients/common/dosemu2.git/test-imagedir/drives/c/io.sys" (cached)
fatfs: full_name: 0 = "/clients/common/dosemu2.git/test-imagedir/drives/c/msdos.sys" (cached)
fatfs: full_name: 0 = "/clients/common/dosemu2.git/test-imagedir/drives/c/boot.blk" (cached)
fatfs: full_name: 0 = "/clients/common/dosemu2.git/test-imagedir/drives/c/io.sys" (cached)
fatfs: system type is "Old MS-DOS (< v4.0)" (0x20000003)                        
fatfs: full_name: 0 = "/clients/common/dosemu2.git/test-imagedir/drives/c/io.sys" (cached)
fatfs: trying to add "/clients/common/dosemu2.git/test-imagedir/drives/c/io.sys":
fatfs: new_obj: created obj 1 (0 left)                                          
fatfs: added as a file                                                          
fatfs: full_name: 0 = "/clients/common/dosemu2.git/test-imagedir/drives/c/msdos.sys" (cached)
fatfs: trying to add "/clients/common/dosemu2.git/test-imagedir/drives/c/msdos.sys":
fatfs: new_obj: created obj 2 (1 left)                                          
fatfs: added as a file                                                          
fatfs: full_name: 0 = "/clients/common/dosemu2.git/test-imagedir/drives/c/command.com" (cached)
fatfs: trying to add "/clients/common/dosemu2.git/test-imagedir/drives/c/command.com":
fatfs: new_obj: created obj 3 (0 left)                                          
fatfs: added as a file                                                          

<snip>
fatfs: new_obj: created obj 8 (1 left)                                          
fatfs: added label "DRIVES C   "                                                
DISK: /clients/common/dosemu2.git/test-imagedir/drives/c: Trying to read 1 sectors at T/S/H 0/0/0+0 at pos 0
fatfs: read: dir /clients/common/dosemu2.git/test-imagedir/drives/c, sec 0, len 1
fatfs: dir /clients/common/dosemu2.git/test-imagedir/drives/c, reading boot sector
############### HELLO txtr->ax == 0xff80 ###################### 
fatfs: made boot block suitable for MS-DOS & PC-DOS, < v4.0                     
INT13: ax=4200 cx=0000 dx=0000                                                  
DISK 00 read [h:0,s:5,t:0](1)->0050:0000                                        
DISK: /clients/common/dosemu2.git/test-imagedir/drives/c: Trying to read 1 sectors at T/S/H 0/5/0+0 at pos 2560
fatfs: read: dir /clients/common/dosemu2.git/test-imagedir/drives/c, sec 5, len 1
fatfs: dir /clients/common/dosemu2.git/test-imagedir/drives/c, reading root dir sector 0
fatfs: reading cluster 0 (+0)                                                   
fatfs: cluster 0 is in object 0                                                 
fatfs: read_dir: obj 0, cluster 0, sec 0                                        
fatfs: assigning up to cluster 0, obj 1                                         
fatfs: assign_clusters: obj 1, start 2, len 8 (io.sys)                          
fatfs: assigning up to cluster 0, obj 2                                         
fatfs: assign_clusters: obj 2, start 10, len 28 (msdos.sys)                     
fatfs: assigning up to cluster 0, obj 3                                         
fatfs: assign_clusters: obj 3, start 38, len 23 (command.com)                   
fatfs: assigning up to cluster 0, obj 4                                         
fatfs: assign_clusters: obj 4, start 61, len 1 (config.sys)                     
fatfs: assigning up to cluster 0, obj 5                                         
fatfs: assign_clusters: obj 5, start 62, len 1 (autoexec.bat)                   
fatfs: assigning up to cluster 0, obj 6                                         
fatfs: scan_dir: reading "dosemu"                                               
fatfs: full_name: 6 = "/clients/common/dosemu2.git/test-imagedir/drives/c/dosemu/"